[WIP] Fix ZeroDivisionWarning in Matern Kernel

MechCoder opened the pull request:
#### Reference Issue

Fixes what @amueller thought was the issue in https://github.com/scikit-learn/scikit-learn/issues/5664,
though that's not really the issue [edited by @amueller so that the issue remains open after this fix is merged]
#### What does this implement/fix? Explain your changes.

The previous code had got around the case where the distance equals zero by setting the gradient to zero after finding where it becomes infinite that raises a `ZeroDivisionWarning`. This is another solution using masked arrays in NumPy. I am unable to reproduce any other warning on master with the latest versions of NumPy and SciPy

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scikit-learn/scikit-learn/7385)

<!-- Reviewable:end -->

MechCoder committed:
Fix ZeroDivisionWarning in Matern Kernel

agramfort commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1320,9 +1320,9 @@ def __call__(self, X, Y=None, eval_gradient=False):
-                K_gradient[~np.isfinite(K_gradient)] = 0
+                dist = np.ma.masked_array(np.sqrt(D.sum(2))[:, :, np.newaxis])
+                K_gradient = K[..., np.newaxis] * D / dist
+                K_gradient = K_gradient.filled(0)
```
K_gradient remains a masked_array afterwards?

MechCoder replied:
no, http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.ma.MaskedArray.filled.html

The notes say "The result is not a masked array"

ogrisel commented:
Can you please add a non regression test with `assert_no_warnings`?

MechCoder committed:
Add NRT using assert_no_warnings

MechCoder commented:
done. hope that was sufficient.

MechCoder committed:
Fix

ogrisel commented on sklearn/gaussian_process/tests/test_kernels.py:
```diff
@@ -317,4 +317,4 @@ def test_repr_kernels():
 
 def test_no_warnings_matern_eval_gradient():
     mat = Matern(length_scale=[0.5, 2.0], nu=0.5)
-    assert_no_warnings(mat, X, eval_gradient=True)
```
You don't need to pass `__call__`. The original line was already correct.

MechCoder committed:
Fix warnings

MechCoder commented:
Sorry about pushing directly. I was not able to reproduce the failure by doing `nosetests-3.4 sklearn/gaussian_process/tests/test_kernels.py ` or `nosetests-3.4 sklearn/gaussian_process/tests`, but I can reproduce while doing `nosetests-3.4 sklearn/`. Just running `nosetests-3.4 sklearn/` seesms to raise plenty of other warnings in `test_kernels.py` though. Trying to investigate.

MechCoder committed:
Remove masked array logic

MechCoder commented:
I was facing some issues in using masked arrays with regards to the warnings. I removed the masked arrays logic and replaced it with a less elegant (but working) solution using initialization of  `K_gradient` with zeros.

These are the following issues that I faced during the usage of masked arrays.

1. In older versions of NumPy doing this, raised a warning. (reported here https://github.com/numpy/numpy/issues/4269).

```python
1 / np.ma.masked_equal([1, 0], 0)
```
It is necessary to do this in those cases.  np.ma.masked_array([1]) /  np.ma.masked_equal([1, 0], 0)

2. Also, seems to be affected by this issue over here (https://github.com/numpy/numpy/issues/4895) while multiplying `K_gradient` and `D`.

I also feel it is necessary to use `a.__call__` explicitly to raise a more meaning full error message here (https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/testing.py#L285) since instances have no `__name__` attribute.

agramfort commented:
LGTM

amueller commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1320,9 +1320,12 @@ def __call__(self, X, Y=None, eval_gradient=False):
-                K_gradient = K[..., np.newaxis] * D \
-                    / np.sqrt(D.sum(2))[:, :, np.newaxis]
-                K_gradient[~np.isfinite(K_gradient)] = 0
+                K_gradient = np.zeros_like(D)
```
Is this faster or slower than the previous solution? Alternatively we could just catch the warning, right?

MechCoder replied:
Yeah, I don't like the solution that I wrote, but I wasn't sure if using the `with warnings.catch_warnings(): warnings.simplefilter` block was encouraged.

amueller replied:
You should catch only the ZeroDivisionError

amueller replied:
Ah. it's RuntimeWarning?

amueller commented:
we use ``np.seterr`` in ``dict_learning``. I think this is probably the best pattern?
No idea how that might interfere with parallel processing?

jnothman commented:
This is marked WIP and has a LGTM. What's the story, @MechCoder? Do we want this in 0.20?

jeremiedbb commented:
The fix was done in https://github.com/scikit-learn/scikit-learn/pull/24245. This PR can be closed.
